### PR TITLE
Feat(QRCode): Implementing barcode in core

### PR DESCRIPTION
**Pull request by MyvTsv**
## Checklist before requesting a review

*Please delete options that are not relevant.*

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.

## Description

Implementing barcode plugin in GLPI's core

### CODE REVIEW

File: templates/generic_show_form.html.twig [hunk 1]
```diff
@@ -683,6 +683,16 @@
                         {{ fields.htmlField('last_boot', item.fields['last_boot']|formatted_datetime(true), __('Last boot date')) }}
                      {% endif %}
 
+
+                     {{  fields.htmlField(
+                        'qrcode',
+                        barcode,
+                        __('Asset URL'),
+                        {
+                           'helper': __('Flash this QRCode to be redirected to the asset page from your mobile device')
```
Comment (cconard96) [inline templates/generic_show_form.html.twig:692]:
```suggestion
                           'helper': __('Scan this QRCode to be redirected to the asset page from your mobile device')
```

File: src/Barcode.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (cconard96) [inline src/Barcode.php:1]: Missing copyright header

File: src/Barcode.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            'black',
+            array(-2, -2, -2, -2)
+        )->setBackgroundColor('white');
+        return $qrcode;
+    }
+
+    public static function renderQRCode(CommonDBTM $item)
+    {
+        global $CFG_GLPI;
+        $lowercaseArray = array_map('strtolower', $CFG_GLPI["asset_types"]);
```
Comment (cconard96) [inline src/Barcode.php:22]: GLPI uses snake_case for variable names.

File: src/Barcode.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        if (
+            in_array(
+                strtolower($item::$rightname),
```
Comment (cconard96) [inline src/Barcode.php:25]: This doesn't seem right. Should just be an in_array check on the asset_types array looking for $item::class?

File: src/Barcode.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+class Barcode extends CommonDropdown
+{
+    public static function generateQRCode(CommonDBTM $item)
+    {
+        $barcode = new \Com\Tecnick\Barcode\Barcode();
+        $qrcode = $barcode->getBarcodeObj(
+            'QRCODE,H',
+            $_SERVER['SERVER_NAME'] . $item->getLinkURL(),
```
Comment (cconard96) [inline src/Barcode.php:10]: Better to use CFG_GLPI['url_base'] instead of SERVER_NAME.

**Review by cconard96 — CHANGES_REQUESTED** [review]

**Comment by stonebuzz** [comment]
can you rebase from `main`  branch

File: templates/generic_show_form.html.twig [hunk 6]
```diff
@@ -683,6 +683,17 @@
...
+                        {{  fields.htmlField(
+                           'qrcode',
+                           barcode,
+                           __('Asset URL'),
+                           {
+                              'helper': __('Scan this QRCode to be redirected to the asset page from your mobile device')
+                           })
+                        }}
+                     {% endif %}
+
```
Comment (stonebuzz) [inline templates/generic_show_form.html.twig:696]:
```suggestion
                     {% if item_type in config("asset_types")  %}
                        {% set barcode %}
                           {{ call('Barcode::renderQRCode', [
                              item
                           ])|raw }}
                        {% endset %}

                        {{  fields.htmlField(
                           'qrcode',
                           barcode,
                           __('Asset URL'),
                           {
                              'helper': __('Scan this QRCode to be redirected to the asset page from your mobile device')
                           })
                        }}
                     {% endif %}

```

Comment (stonebuzz) [reply templates/generic_show_form.html.twig:696]: The CFG_GLPI[“asset_types”] constant (give from twig function `config`)  is used to determine whether or not to display the QRCode.

this leaves the asset “showForm” unchanged

File: src/Certificate.php [hunk 7]
```diff
@@ -550,6 +550,7 @@ public function showForm($ID, array $options = [])
         TemplateRenderer::getInstance()->display('pages/management/certificate.html.twig', [
             'item'   => $this,
             'params' => $options,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/Certificate.php:553]:
```suggestion
```

use less now (see my last comment)

File: src/CommonDBTM.php [hunk 8]
```diff
@@ -484,6 +484,7 @@ public function showForm($ID, array $options = [])
             'params' => $options,
             'no_header' => !$new_item && !$in_modal,
             'cluster' => $cluster,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/CommonDBTM.php:487]:
```suggestion
```

use less now (see my last comment)

File: src/Monitor.php [hunk 9]
```diff
@@ -155,6 +155,7 @@ public function showForm($ID, array $options = [])
         TemplateRenderer::getInstance()->display('pages/assets/monitor.html.twig', [
             'item'   => $this,
             'params' => $options,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/Monitor.php:158]:
```suggestion
```

use less now (see my last comment)

File: src/Phone.php [hunk 10]
```diff
@@ -175,6 +175,7 @@ public function showForm($ID, array $options = [])
         TemplateRenderer::getInstance()->display('pages/assets/phone.html.twig', [
             'item'   => $this,
             'params' => $options,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/Phone.php:178]:
```suggestion
```

use less now (see my last comment)

File: src/Printer.php [hunk 11]
```diff
@@ -301,6 +301,7 @@ public function showForm($ID, array $options = [])
         TemplateRenderer::getInstance()->display('pages/assets/printer.html.twig', [
             'item'   => $this,
             'params' => $options,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/Printer.php:304]:
```suggestion
```

use less now (see my last comment)

File: src/SoftwareLicense.php [hunk 12]
```diff
@@ -274,6 +274,7 @@ public function showForm($ID, array $options = [])
         TemplateRenderer::getInstance()->display('pages/management/softwarelicense.html.twig', [
             'item'   => $this,
             'params' => $options,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/SoftwareLicense.php:277]:
```suggestion
```

use less now (see my last comment)

File: src/Unmanaged.php [hunk 13]
```diff
@@ -85,6 +85,7 @@ public function showForm($ID, array $options = [])
         TemplateRenderer::getInstance()->display('pages/assets/unmanaged.html.twig', [
             'item'   => $this,
             'params' => $options,
+            'barcode' => Barcode::renderQRCode($this)
```
Comment (stonebuzz) [inline src/Unmanaged.php:88]:
```suggestion
```

use less now (see my last comment)

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Config.php [hunk 14]
```diff
@@ -1597,6 +1597,9 @@ public static function checkExtensions($list = null)
                 'simplexml' => [
                     'required'  => true,
                 ],
+                'bcmath' => [
```
Comment (cedric-anne) [inline src/Config.php:1600]: If the extension is required, please add it to the composer requirements too (`composer require ext-bcmath`).

Comment (stonebuzz) [reply src/Config.php:1600]: @cedric-anne 
this https://github.com/glpi-project/docker-images/pull/131 is still needed right?

Comment (trasher) [reply src/Config.php:1600]: yes

**Review by stonebuzz — APPROVED** [review]

**Comment by cconard96** [comment]
Not strictly related to this PR, but last year there was a quick discussion about integrating the `camerainput` plugin into the core. In my mind, it is strongly related to the old barcode plugin and thought both plugins could be integrated at the same time (different PRs probably).

The barcode plugin allowed creating barcodes/qr codes while the camerainput plugin lets you use a regular camera/webcam to scan barcode/qrcode data directly into specific fields.

https://github.com/cconard96/glpi-camerainput-plugin

Not sure how everyone else feels about it, but I wanted to bring this up now before I forget about it again.

**Review by cconard96 — APPROVED** [review]

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]

**Comment by AdrienClairembault** [comment]
Can you rebase and fix the conflicts ?

**Review by AdrienClairembault — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**